File: watch/wearos/src/main/java/app/vescape/wear/NavRoute.kt
package app.vescape.wear

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.unit.dp

/**
 * Route polyline under the gauges, drawn heading-up with the rider pinned near the watch centre.
 * Sits at the bottom of the frame's layer stack so gauges, readouts and the nav chevron draw over it.
 *
 * One source: the real polyline the phone pushed on [ROUTE_PATH], placed by the frame's rider lanes.
 * Until those arrive — a route pushed but no fix yet — only the rider dot draws, so the wrist never
 * shows a line the rider is not actually on.
 */
@Composable
internal fun NavRoute(frame: WatchFrame, muted: Boolean, navFocus: () -> Float = { 0f }) {
    val color = if (muted) DimText else navColor()
    val route = RouteState.route.value
    val east = frame.riderEastM
    val north = frame.riderNorthM
    val targetRouteSpanM = (frame.routeSpanM ?: DEFAULT_ROUTE_SPAN_M).toFloat()
        .coerceIn(MIN_ROUTE_SPAN_M, MAX_ROUTE_SPAN_M)
    val routeSpanM by animateFloatAsState(
        targetValue = targetRouteSpanM,
        animationSpec = tween(durationMillis = ROUTE_ZOOM_EASE_MS, easing = FastOutSlowInEasing),
        label = "routeZoom",
    )

    if (route != null && east != null && north != null) {
        AnimatedRoute(
            route = route,
            targetEastM = east.toFloat(),
            targetNorthM = north.toFloat(),
            targetCourseDeg = (frame.courseDeg ?: 0.0).toFloat(),
            routeSpanM = routeSpanM,
            color = color,
            navFocus = navFocus,
        )
    }

    Canvas(modifier = Modifier.fillMaxSize()) {
        // Rider sits below the screen centre so more of the frame is "ahead" than behind.
        drawRiderDot(Offset(size.width / 2f, size.height / 2f + RIDER_DROP.toPx()), color)
    }
}

/**
 * "You are here": a ring in the route's own colour, punched out to black so whatever passes under
 * it never reads as passing through the rider. Shared with the radar page — the two are the same
 * map seen at different scales, and a rider that looked different on each would say otherwise.
 */
internal fun DrawScope.drawRiderDot(center: Offset, color: Color) {
    drawCircle(Color.Black, radius = RIDER_DOT_R.toPx(), center = center)
    drawCircle(
        color,
        radius = RIDER_DOT_R.toPx(),
        center = center,
        style = Stroke(width = RIDER_RING_W.toPx()),
    )
}

@Composable
private fun AnimatedRoute(
    route: WatchRoute,
    targetEastM: Float,
    targetNorthM: Float,
    targetCourseDeg: Float,
    routeSpanM: Float,
    color: Color,
    navFocus: () -> Float,
) {
    // Offsets are metres from *this* route's origin. A new route moves the origin, so the old
    // animators would glide the rider across a jump that never happened: key them to the route.
    val eastM = remember(route) { Animatable(targetEastM) }
    val northM = remember(route) { Animatable(targetNorthM) }
    val courseDeg = remember(route) { Animatable(targetCourseDeg) }
    val motionSpec = tween<Float>(durationMillis = ROUTE_MOTION_EASE_MS, easing = LinearEasing)

    LaunchedEffect(targetEastM) { eastM.animateTo(targetEastM, motionSpec) }
    LaunchedEffect(targetNorthM) { northM.animateTo(targetNorthM, motionSpec) }
    LaunchedEffect(targetCourseDeg) {
        courseDeg.animateTo(
            courseDeg.value + shortestAngleDelta(courseDeg.value, targetCourseDeg),
            motionSpec,
        )
    }

    // A route runs for kilometres and Compose does not clip by default, so without this the line
    // reaches past the frame and draws over whatever page sits next to it. On a round watch the
    // bounds are still square, so it stops inside the gauge ring instead: the line belongs under the
    // gauges, never crossing them out to the bezel.
    val isRound = LocalConfiguration.current.isScreenRound
    Canvas(modifier = Modifier.fillMaxSize()) {
        // Read in the draw scope: the line thickens as the map page takes over, without
        // recomposing anything. Only there — a wider line under the readouts would fight the
        // gauge fills, but on the map it is the whole page and has to survive sunlight.
        val focus = navFocus().coerceIn(0f, 1f)
        val center = Offset(size.width / 2f, size.height / 2f + RIDER_DROP.toPx())
        val scale = (size.minDimension - ROUTE_EDGE_INSET.toPx()) / routeSpanM
        val path = routePath(route, Offset(eastM.value, northM.value), center, scale)
        val faceCenter = Offset(size.width / 2f, size.height / 2f)
        // One pixel inside the gauge circle's guide line, so the route stops just short of it.
        val faceRadius = size.minDimension / 2f - GAUGE_RIM_INSET.toPx() - GUIDE_HALF_WIDTH.toPx() - 1f
        val faceClip = Path().apply {
            if (isRound) {
                addOval(Rect(faceCenter, faceRadius))
            } else {
                addRect(Rect(Offset.Zero, size))
            }
        }
        clipPath(faceClip) {
            // Heading-up, taking the shortest turn across the 0°/360° boundary.
            rotate(degrees = -courseDeg.value, pivot = center) {
                drawRoute(path, color, focus)
            }
        }
    }
}

internal fun shortestAngleDelta(fromDeg: Float, toDeg: Float): Float =
    (((toDeg - fromDeg + 180f) % 360f + 360f) % 360f) - 180f

/** Route points (metres east/north of the route origin) into screen pixels around the rider. */
private fun routePath(route: WatchRoute, rider: Offset, center: Offset, scale: Float): Path =
    polyline(
        route.points.map { point ->
            Offset(
                center.x + (point.eastM - rider.x) * scale,
                // Screen y grows downward; north is up.
                center.y - (point.northM - rider.y) * scale,
            )
        },
    )

/** One flat stroke: a casing glow under the line only muddied it against the gauge fills. */
private fun DrawScope.drawRoute(path: Path, color: Color, focus: Float) {
    val width = ROUTE_W.toPx() + (ROUTE_FOCUS_W - ROUTE_W).toPx() * focus
    val alpha = ROUTE_ALPHA + (ROUTE_FOCUS_ALPHA - ROUTE_ALPHA) * focus
    drawPath(path, color.copy(alpha = alpha), style = routeStroke(width))
}

private fun polyline(points: List<Offset>): Path = Path().apply {
    moveTo(points[0].x, points[0].y)
    for (index in 1 until points.size) lineTo(points[index].x, points[index].y)
}

private fun routeStroke(width: Float) = Stroke(width = width, cap = StrokeCap.Round, join = StrokeJoin.Round)

/** Fallback metres of route across the watch face until the phone publishes its camera span. */
private const val DEFAULT_ROUTE_SPAN_M = 600.0
private const val MIN_ROUTE_SPAN_M = 150f
private const val MAX_ROUTE_SPAN_M = 2_000f
private const val ROUTE_ZOOM_EASE_MS = 350
private const val ROUTE_MOTION_EASE_MS = 300

private val ROUTE_EDGE_INSET = 24.dp
/** Half the widest gauge guide stroke: the route clip stops at the inner side of that line. */
private val GUIDE_HALF_WIDTH = 1.dp
private val ROUTE_W = 2.dp

/** Width and opacity on the map page, where the line is the page and has to read in daylight. */
private val ROUTE_FOCUS_W = 3.5.dp
private const val ROUTE_ALPHA = 0.55f
private const val ROUTE_FOCUS_ALPHA = 0.85f
private val RIDER_DOT_R = 4.dp
/** Same weight as the route line, so the rider reads as part of it rather than an added marker. */
private val RIDER_RING_W = ROUTE_W
private val RIDER_DROP = 34.dp

File: watch/watchos/NavRoute.swift
import SwiftUI

/// Route polyline under the gauges, drawn heading-up with the rider pinned near the watch centre.
/// Sits at the bottom of the frame's layer stack so gauges, readouts and the nav chevron draw over
/// it.
///
/// One source: the real polyline the phone pushed on the route channel, placed by the frame's rider
/// lanes. Until those arrive — a route pushed but no fix yet — only the rider dot draws, so the
/// wrist never shows a line the rider is not actually on.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/NavRoute.kt `NavRoute`
/// @platform-diff Wear OS clips the line to the circle its gauges ring, because a round panel's
///   drawing bounds are square and the line would otherwise run out to the bezel. Here the clip is
///   the display's own rounded rectangle, one step inside the rim gauges (``Rim``), so the route
///   uses the corners the rectangle actually has instead of the circle it does not.
struct NavRoute: View {
  let route: WatchRoute?
  /// Bumped by ``PhoneLink`` on every route change, so the motion animators reset with the origin
  /// they are measured against rather than gliding across a jump that never happened.
  let generation: Int
  let frame: WatchFrame
  /// Nav-focus progress: the line thickens and brightens as the nav page takes the screen, where it
  /// is the whole page and has to read in daylight.
  var focus: Double = 0
  /// The nav accent already resolved by the caller — the rider's colour, dimmed when the frame is
  /// stale. Resolved outside so the route, the chevron and the rider dot cannot disagree.
  var color: Color

  /// Course as an unwrapped angle, so a heading crossing north turns the short way instead of
  /// spinning 359° back around the compass.
  @State private var unwrappedCourse: Double = 0

  var body: some View {
    GeometryReader { geometry in
      let centre = CGPoint(
        x: geometry.size.width / 2,
        y: geometry.size.height / 2 + RIDER_DROP
      )
      ZStack {
        if let route, let east = frame.riderEastM, let north = frame.riderNorthM {
          RouteShape(
            route: route,
            eastM: east,
            northM: north,
            courseDeg: unwrappedCourse,
            spanM: clampedSpan
          )
          .stroke(
            color.opacity(ROUTE_ALPHA + (ROUTE_FOCUS_ALPHA - ROUTE_ALPHA) * focus),
            style: StrokeStyle(
              lineWidth: ROUTE_WIDTH + (ROUTE_FOCUS_WIDTH - ROUTE_WIDTH) * focus,
              lineCap: .round,
              lineJoin: .round
            )
          )
          // A route runs for kilometres; without this it reaches past the display and draws over
          // the rim gauges. One step inside them, so the line stops just short of the gauge lines.
          .clipShape(Rim.path(in: geometry.size, inset: Rim.innerInset))
          .animation(.linear(duration: ROUTE_MOTION_EASE), value: east)
          .animation(.linear(duration: ROUTE_MOTION_EASE), value: north)
          .animation(.linear(duration: ROUTE_MOTION_EASE), value: unwrappedCourse)
          .animation(.easeInOut(duration: ROUTE_ZOOM_EASE), value: clampedSpan)
          // Offsets are metres from *this* route's origin. A new route moves the origin, so the
          // animators would glide the rider across a jump that never happened: a new generation is
          // a new view identity, which starts them from the new route's own numbers.
          .id(generation)
        }
        Canvas { context, _ in context.drawRiderDot(at: centre, color: color) }
      }
      .onChange(of: frame.courseDeg ?? 0) { _, next in
        unwrappedCourse += shortestAngleDelta(from: unwrappedCourse, to: next)
      }
      .onAppear { unwrappedCourse = frame.courseDeg ?? 0 }
    }
  }

  private var clampedSpan: Double {
    min(MAX_ROUTE_SPAN_M, max(MIN_ROUTE_SPAN_M, frame.routeSpanM ?? DEFAULT_ROUTE_SPAN_M))
  }
}

/// "You are here": a ring in the route's own colour, punched out to black so whatever passes under
/// it never reads as passing through the rider.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/NavRoute.kt `drawRiderDot`
extension GraphicsContext {
  func drawRiderDot(at centre: CGPoint, color: Color) {
    let circle = Path(
      ellipseIn: CGRect(
        x: centre.x - RIDER_DOT_RADIUS, y: centre.y - RIDER_DOT_RADIUS,
        width: RIDER_DOT_RADIUS * 2, height: RIDER_DOT_RADIUS * 2
      )
    )
    fill(circle, with: .color(.black))
    stroke(circle, with: .color(color), lineWidth: ROUTE_WIDTH)
  }
}

/// The polyline in screen space: route points are metres east/north of the route origin, placed
/// around the rider and rotated heading-up.
///
/// A `Shape` rather than a `Canvas` for one reason: the rider's position, the course and the zoom
/// all move continuously, and a shape's `animatableData` is what interpolates them between frames.
/// Drawing the same geometry in a canvas would jump once per push.
private struct RouteShape: Shape {
  let route: WatchRoute
  var eastM: Double
  var northM: Double
  var courseDeg: Double
  var spanM: Double

  var animatableData: AnimatablePair<AnimatablePair<Double, Double>, AnimatablePair<Double, Double>> {
    get { AnimatablePair(AnimatablePair(eastM, northM), AnimatablePair(courseDeg, spanM)) }
    set {
      eastM = newValue.first.first
      northM = newValue.first.second
      courseDeg = newValue.second.first
      spanM = newValue.second.second
    }
  }

  func path(in rect: CGRect) -> Path {
    guard route.points.count > 1, spanM > 0 else { return Path() }
    // Rider sits below the centre so more of the display is "ahead" than behind.
    let centre = CGPoint(x: rect.midX, y: rect.midY + RIDER_DROP)
    let scale = (min(rect.width, rect.height) - ROUTE_EDGE_INSET) / spanM

    var path = Path()
    for (index, point) in route.points.enumerated() {
      let screen = CGPoint(
        x: centre.x + (point.eastM - eastM) * scale,
        // Screen y grows downward; north is up.
        y: centre.y - (point.northM - northM) * scale
      )
      if index == 0 { path.move(to: screen) } else { path.addLine(to: screen) }
    }
    // Heading-up: the world rotates under a rider who always points at the top of the display.
    return path.applying(
      CGAffineTransform(translationX: centre.x, y: centre.y)
        .rotated(by: -courseDeg * .pi / 180)
        .translatedBy(x: -centre.x, y: -centre.y)
    )
  }
}

/// Shortest turn between two compass headings, in degrees, signed.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/NavRoute.kt `shortestAngleDelta`
func shortestAngleDelta(from: Double, to: Double) -> Double {
  (((to - from + 180).truncatingRemainder(dividingBy: 360) + 360).truncatingRemainder(dividingBy: 360)) - 180
}

/// Fallback metres of route across the display until the phone publishes its camera span.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/NavRoute.kt `DEFAULT_ROUTE_SPAN_M`
private let DEFAULT_ROUTE_SPAN_M = 600.0
private let MIN_ROUTE_SPAN_M = 150.0
private let MAX_ROUTE_SPAN_M = 2_000.0
private let ROUTE_ZOOM_EASE = 0.35
private let ROUTE_MOTION_EASE = 0.3

private let ROUTE_EDGE_INSET: CGFloat = 24
private let ROUTE_WIDTH: CGFloat = 2
/// Width and opacity on the nav page, where the line is the page and has to read in daylight.
private let ROUTE_FOCUS_WIDTH: CGFloat = 3.5
private let ROUTE_ALPHA = 0.55
private let ROUTE_FOCUS_ALPHA = 0.85
private let RIDER_DOT_RADIUS: CGFloat = 4
private let RIDER_DROP: CGFloat = 34
